import_export/admin.py

```diff
@@ -177,34 +191,182 @@ def get_import_context_data(self, **kwargs):
+
+    def get_import_form_kwargs(self, request):
+        """
+        .. versionadded:: 3.0
+
+        Return a dictionary of values with which to initialize the 'import'
+        form (including the initial values returned by
+        :meth:`~import_export.admin.ImportMixin.get_import_form_initial`).
+        """
+        return {
+            "data": request.POST or None,
+            "files": request.FILES or None,
+            "initial": self.get_import_form_initial(request),
+        }
+
+    def get_import_form_initial(self, request):
+        """
+        .. versionadded:: 3.0
+
+        Return a dictionary of initial field values to be provided to the
+        'import' form.
+        """
+        return {}
+
+    def create_confirm_form(self, request, import_form=None):
+        """
+        .. versionadded:: 3.0
+
+        Return a form instance to use for the 'confirm' import step.
+        This method can be extended to make dynamic form updates to the
+        form after it has been instantiated. You might also look to
+        override the following:
+
+        * :meth:`~import_export.admin.ImportMixin.get_confirm_form_class`
+        * :meth:`~import_export.admin.ImportMixin.get_confirm_form_kwargs`
+        * :meth:`~import_export.admin.ImportMixin.get_confirm_form_initial`
+        """
+        form_class = self.get_confirm_form_class(request)
+        kwargs = self.get_confirm_form_kwargs(request, import_form)
+        return form_class(**kwargs)
+
+    def get_confirm_form_class(self, request):
+        """
+        .. versionadded:: 3.0
+
+        Return the form class to use for the 'confirm' import step. If you only
+        have a single custom form class, you can set the ``confirm_form_class``
+        attribute to change this for your subclass.
+        """
+        # TODO: Remove following conditional when get_confirm_import_form() is removed
+        if not getattr(self.get_confirm_import_form, 'is_original', False):
+            warnings.warn(
+                "ImportMixin.get_confirm_import_form() is deprecated and will "
+                "be removed in a future release. Please use the new "
+                "'confirm_form_class' attribute to specify a custom form "
+                "class, or override the get_confirm_form_class() method if "
+                "your requirements are more complex.",
+                category=DeprecationWarning
+            )
+            return self.get_confirm_import_form()
+        # Return the class attribute value
+        return self.confirm_form_class
+
+    def get_confirm_form_kwargs(self, request, import_form=None):
+        """
+        .. versionadded:: 3.0
+
+        Return a dictionary of values with which to initialize the 'confirm'
+        form (including the initial values returned by
+        :meth:`~import_export.admin.ImportMixin.get_confirm_form_initial`).
+        """
+        return {
+            "data": request.POST or None,
```

Comment:
Seems like there is an issue here when importing via the browser.  This process doesn't work when called from [`import_action()`](https://github.com/django-import-export/django-import-export/pull/1147/files#diff-a6dabb2139e7ec78138476e9a12311c22cc93cbeab4a93994d4f95f64934805eR468).  The `data` dict has to be empty.
However when called from `process_import()`, the data dict must be passed.

Reply:
Ah, I see... In `import_action()` the posted data is for the import form, not the confirm form. Thanks for the explanation!

Reply:
Thanks Andy.  Import works fine now via the console.  We have a failing test which I will look into. (annoyingly tox on CI doesn't report this as an error).

Reply:
I have tracked down the source of this error (broken implementation of `CustomBookAdmin.get_confirm_form_initial()` and fixed it.